### PR TITLE
NFC: Amend ndef parsing design.

### PR DESCRIPTION
**Pull request by pan-**
* Decouple Message parsing from Record parsing
* Decouple record parsing from Common objects
* Provide an easy to use parser that parse the most common objects.

### CODE REVIEW

File: features/nfc/doc/nfc_design.md [hunk 1]
```diff
@@ -415,10 +506,15 @@ GreenTea tests will be provided to partners to ensure compliance with the NFC EE
 * Event Queue
 
 There are currently at least four event queues (Plaftorm, BLE, USB, IP) in mbed OS and NFC will also require an event queing mechanism. We should aim at reusing one of these existing queues with the long term goal of unifying these code bases.
-					
+
 [phase_1_architecture]: phase_1_architecture.png
 [phase_2_architecture]: phase_2_architecture.png
 [nfc_controller_diagram]: uml_diagram_controller.png
 [nfc_endpoints_diagram]: uml_diagram_endpoints.png
 [ndef_diagram]: uml_diagram_ndef.png
```
Comment (donatieng) [inline features/nfc/doc/nfc_design.md]: This line will need to be removed

File: features/nfc/doc/nfc_design.md [hunk 2]
```diff
@@ -261,99 +261,190 @@ The `NFCRemoteTarget` class derives from `NFCTarget` and additionally from `NFCE
...
-#### NDEF Message builder
+#### ndef::MessageParser
 
-We're using a builder pattern to encode an NDEF message over a byte array.
+![ndef_message_parser_diagram]
+
+Messages incoming from the peer are parsed by a `MessageParser` which produce
+`Record` instances to its client. The parsing operation is event-driven: a
+message parser client register a delegate inside the message parser. This delegate
+get notified whenever an interesting event happen during the parsing.
```
Comment (donatieng) [inline features/nfc/doc/nfc_design.md]: --> The parsing operation is event-driven: a
 message parser client register**s** a delegate inside the message parser. This delegate
 get**s** notified whenever an interesting event happen**s** during the parsing.

**Review by donatieng — CHANGES_REQUESTED** [review]
Thanks @pan-, two minimal formatting changes needed but looks good!

File: features/nfc/doc/nfc_design.md (hunk 2 above)
Comment (pan-) [reply features/nfc/doc/nfc_design.md]: I wonder I never apply the correct form. It is not that difficult. 
Thanks.

**Comment by pan-** [comment]
@donatieng I think I've addressed your comments.

**Pull request merged**